raiden/network/rpc/smartcontract_proxy.py

```diff
@@ -181,12 +182,30 @@ def decode_event(self, log: Dict):
     def encode_function_call(self, function: str, args: List = None):
         return self.get_transaction_data(self.contract.abi, function, args)
 
-    def estimate_gas(self, function: str, *args):
+    def estimate_gas(self, block_identifier, function: str, *args) -> typing.Optional[int]:
+        """Returns a gas estimate for the function with the given arguments or
+        None if the function call will fail due to Insufficient funds or
+        the logic in the called function."""
         fn = getattr(self.contract.functions, function)
+        address = to_checksum_address(self.jsonrpc_client.address)
+        if self.jsonrpc_client.eth_node == constants.EthClient.GETH:
+            # Unfortunately geth does not follow the ethereum JSON-RPC spec and
+            # does not accept a block identifier argument for eth_estimateGas
+            # parity and py-evm (trinity) do.
+            #
+            # Geth only runs estimateGas on the pending block and that's why we
+            # should also enforce parity, py-evm and others to do the same since
+            # we can't customize geth.
+            #
+            # Spec: https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_estimategas
+            # Geth Issue: https://github.com/ethereum/go-ethereum/issues/2586
+            # Relevant web3 PR: https://github.com/ethereum/web3.py/pull/1046
+            block_identifier = None
```

Comment:
should we add an assert for `block_identifer == 'pending'` because of this?

Reply:
yes. I actually was thinking of doing this (until geth also implements the option to set a `block_identifier`) ... but did not do this yet. Thanks for reminding.